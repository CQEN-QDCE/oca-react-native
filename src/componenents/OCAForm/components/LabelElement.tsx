import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import type { LabelElementProps } from '../../types';

export const LabelElement = ({ attribute, style }: LabelElementProps) => {
  return (
    <View>
      <Text style={[styles.label, style]} numberOfLines={1}>
        {attribute.name}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  label: {
    color: '#333',
    fontWeight: 'bold',
    fontSize: 14,
  },
});
