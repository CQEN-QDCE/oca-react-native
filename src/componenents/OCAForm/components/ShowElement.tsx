import React from 'react';
import { TouchableOpacity, View, Text } from 'react-native';
import type { ShowElementProps } from '../../types';
import { attributeTypeStyle } from '../../utils/attributeTypeStyle';

export const ShowElement = ({
  hideShowOptions,
  onToggleViewPressed,
  shown = false,
  style,
}: ShowElementProps) => {
  return (
    <View>
      {hideShowOptions.visibility && (
        <TouchableOpacity activeOpacity={1} onPress={onToggleViewPressed}>
          <View>
            {shown ? (
              hideShowOptions?.labelHide ? (
                hideShowOptions.labelHide
              ) : (
                <Text style={[attributeTypeStyle.text, style]}>Hide</Text>
              )
            ) : hideShowOptions?.labelShow ? (
              hideShowOptions.labelShow
            ) : (
              <Text style={[attributeTypeStyle.text, style]}>Show</Text>
            )}
          </View>
        </TouchableOpacity>
      )}
    </View>
  );
};
