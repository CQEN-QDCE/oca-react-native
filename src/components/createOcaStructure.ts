import type { OCA } from 'oca.js';
import { OcaJs } from '../packages/oca.js-form-core/OcaJs';

export const createOcaStructure = async (oca?: OCA) => {
  const ocaJs = new OcaJs({});
  if (oca) {
    return await ocaJs.createStructure(oca);
  }
  return undefined;
};
