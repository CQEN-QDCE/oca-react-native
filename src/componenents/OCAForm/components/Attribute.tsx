import React from 'react';
import type { AttributeProps } from '../types';
import { StyleSheet, View } from 'react-native';
import { LabelElement } from './LabelElement';
import { ShowElement } from './ShowElement';
import { TextAttribute } from './AttributesTypes/TextAttribute';
import { DateAttribute } from './AttributesTypes/DateAttribute';
import { SelectAttribute } from './AttributesTypes/SelectAttribute';
import { BinaryAttribute } from './AttributesTypes/BinaryAttribute';

export const Attribute = ({
  attribute,
  shown,
  hideShowOptions,
  onToggleViewPressed,
  stylingOptions,
  attributeLabel = null,
  attributeValue = null,
}: AttributeProps) => {
  const attributeTypeSelector = () => {
    if (attribute.type === 'Date') {
      console.log('DATE VALUE: ' + attribute.name + ' - ' + attribute.value);
      return (
        <DateAttribute
          attribute={attribute}
          shown={shown}
          styles={stylingOptions?.textStyle}
        />
      );
    } else if (attribute.type === 'Select') {
      return (
        <SelectAttribute
          attribute={attribute}
          shown={shown}
          styles={stylingOptions?.textStyle}
        />
      );
    } else if (attribute.type === 'Binary') {
      return (
        <BinaryAttribute
          attribute={attribute}
          shown={shown}
          binaryStyle={stylingOptions?.binaryStyle}
          textStyle={stylingOptions?.textStyle}
        />
      );
    }
    return (
      <TextAttribute
        attribute={attribute}
        shown={shown}
        styles={stylingOptions?.textStyle}
      />
    );
  };

  return (
    <View style={[styles.container, stylingOptions?.attributeContainerStyle]}>
      {attributeLabel ? (
        attributeLabel(attribute)
      ) : (
        <LabelElement
          attribute={attribute}
          style={stylingOptions?.labelTextStyle}
        />
      )}

      {attributeValue ? (
        attributeValue(attribute)
      ) : (
        <View style={styles.valueContainer}>
          {attributeTypeSelector()}

          <ShowElement
            hideShowOptions={hideShowOptions}
            shown={shown}
            onToggleViewPressed={onToggleViewPressed}
          />
        </View>
      )}
      <View style={[styles.separator, stylingOptions?.separatorStyle]} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    paddingHorizontal: 25,
    paddingTop: 16,
    backgroundColor: 'transparent',
  },
  valueContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 10,
    maxWidth: '100%',
  },
  separator: {
    borderBottomColor: '#c9c9c9',
    borderBottomWidth: 2,
    paddingTop: 12,
  },
});
