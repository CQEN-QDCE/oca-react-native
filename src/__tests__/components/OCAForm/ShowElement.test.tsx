import React from 'react';
import { StyleSheet } from 'react-native';
import { ShowElement } from '../../../components/OCAForm/components/ShowElement';
import { Text } from 'react-native';
import { render } from '@testing-library/react-native';
const onToggleViewPressed = () => {};

describe('ShowElement', () => {
  it('Given visible hideShowOptions and visible attribute value should render the hide button properly', async () => {
    const { findByText } = render(
      <ShowElement
        hideShowOptions={{ visibility: true }}
        onToggleViewPressed={onToggleViewPressed}
        shown={true}
      />
    );

    const hideButton = await findByText('Hide');

    expect(hideButton).toBeTruthy();
  });

  it('Given visible hideShowOptions and hidden attribute value should render the show button properly', async () => {
    const { findByText } = render(
      <ShowElement
        hideShowOptions={{ visibility: true }}
        onToggleViewPressed={onToggleViewPressed}
        shown={false}
      />
    );

    const showButton = await findByText('Show');
    expect(showButton).toBeTruthy();
  });

  it('Given a hidden hideShowOptions should not display the hide/show button', () => {
    const tree = (
      <ShowElement
        hideShowOptions={{ visibility: false }}
        onToggleViewPressed={onToggleViewPressed}
        shown={false}
      />
    );

    expect(tree).toMatchSnapshot();
  });

  it('Given a visible hideShowOptions, a hidden attribute value and a custom show label should render the custom show label', async () => {
    const { findByText } = render(
      <ShowElement
        hideShowOptions={{
          visibility: true,
          labelShow: <Text>Afficher</Text>,
        }}
        onToggleViewPressed={onToggleViewPressed}
        shown={false}
      />
    );

    const customShowButton = await findByText('Afficher');
    expect(customShowButton).toBeTruthy();
  });

  it('Given a visible hideShowOptions, a visible attribute value and a custom hide label should render the custom hide label', () => {
    const { findByText } = render(
      <ShowElement
        hideShowOptions={{
          visibility: true,
          labelHide: <Text>Cacher</Text>,
        }}
        onToggleViewPressed={onToggleViewPressed}
        shown={true}
      />
    );

    const customHideButton = findByText('Cacher');
    expect(customHideButton).toBeTruthy();
  });

  it('Given a visible hideShowOptions and a custom styling should render ShowElement with the style', () => {
    const styles = StyleSheet.create({
      hideShow: {
        fontSize: 20,
        textDecorationLine: 'underline',
      },
    });
    const tree = (
      <ShowElement
        hideShowOptions={{
          visibility: true,
        }}
        onToggleViewPressed={onToggleViewPressed}
        shown={true}
        style={styles.hideShow}
      />
    );

    expect(tree).toMatchSnapshot();
  });
});
