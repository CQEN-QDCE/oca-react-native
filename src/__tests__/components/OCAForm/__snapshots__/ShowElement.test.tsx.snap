// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowElement Given a hidden hideShowOptions should not display the hide/show button 1`] = `
<ShowElement
  hideShowOptions={
    Object {
      "visibility": false,
    }
  }
  onToggleViewPressed={[Function]}
  shown={false}
/>
`;

exports[`ShowElement Given a visible hideShowOptions and a custom styling should render ShowElement with the style 1`] = `
<ShowElement
  hideShowOptions={
    Object {
      "visibility": true,
    }
  }
  onToggleViewPressed={[Function]}
  shown={true}
  style={
    Object {
      "fontSize": 20,
      "textDecorationLine": "underline",
    }
  }
/>
`;
